src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySinkConfig.java

```diff
@@ -172,6 +187,79 @@ void validateCmekKey(FailureCollector failureCollector, Map<String, String> argu
     validateCmekKeyLocation(cmekKeyName, null, location, failureCollector);
   }
 
+  /**
+   * Validates job label key value pairs, as per the following rules:
+   * Keys and values can contain only lowercase letters, numeric characters, underscores, and dashes.
+   * Defined in the following link:
+   * <a href="https://cloud.google.com/bigquery/docs/adding-labels#job-label">Docs</a>
+   * @param failureCollector failure collector
+   */
+  void validateJobLabelKeyValue(FailureCollector failureCollector) {
```

Comment:
I don't see the validation for 1st and 3rd bullet points here: https://cloud.google.com/bigquery/docs/labels-intro#requirements

Am I missing something?

Reply:
Yes, i have missed this i was following validation as per this [doc](https://cloud.google.com/bigquery/docs/adding-labels#adding_a_label_without_a_value).

Updated !

```java
Set<String> reservedKeys = BigQueryUtil.BQ_JOB_LABEL_SYSTEM_KEYS;
int MAX_LABELS = 64 - reservedKeys.size();
int MAX_KEY_LENGTH = 63;
int MAX_VALUE_LENGTH = 63;

// ...code... //

// Key cannot be longer than 63 characters
if (key.length() > MAX_KEY_LENGTH) {
 failureCollector.addFailure(String.format("Invalid job label key '%s'.", key),
                "Job label key cannot be longer than 63 characters.").withConfigProperty(NAME_BQ_JOB_LABELS);
 isKeyValid = false;
}

// Value cannot be longer than 63 characters
if (value.length() > MAX_VALUE_LENGTH) {
  failureCollector.addFailure(String.format("Invalid job label value '%s'.", value),
                "Job label value cannot be longer than 63 characters.").withConfigProperty(NAME_BQ_JOB_LABELS);
isValueValid = false;
}

// ...code... //

// Check if number of labels is greater than 64 - reserved keys
if (uniqueKeys.size() > MAX_LABELS) {
      failureCollector.addFailure("Number of job labels exceeds the limit.",
              String.format("Number of job labels cannot be greater than %d.", MAX_LABELS))
              .withConfigProperty(NAME_BQ_JOB_LABELS);
}
```

Reply:
I have added new test case for these validations.
```java
public void testJobLabelWith65Keys() {  // ...code ... // }
public void testJobLabelWithKeyLength64() {  // ...code ... // }
public void testJobLabelWithValueLength64() {  // ...code ... // }
```

Reply:
Thanks.